Next.js: Add @storybook/experimental-nextjs-vite package

valentinpalkovic opened the pull request:
Closes https://github.com/storybookjs/storybook/issues/28799

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

I have created a new experimental `@storybook/experimental-nextjs-vite` package, based on Vite and `vite-plugin-storybook-nextjs`

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Run the experimental-nextjs-vite sandbox
2. Open Storybook in your browser
3. Make sure Storybook works

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-28800-sha-2528064a`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-28800-sha-2528064a sandbox` or in an existing project with `npx storybook@0.0.0-pr-28800-sha-2528064a upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-28800-sha-2528064a`](https://npmjs.com/package/storybook/v/0.0.0-pr-28800-sha-2528064a) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`valentin/nextjs-vite`](https://github.com/storybookjs/storybook/tree/valentin/nextjs-vite) |
| **Commit** | [`2528064a`](https://github.com/storybookjs/storybook/commit/2528064a2f9457aa2b835c40b93b6c6b7de6ecb4) |
| **Datetime** | Sat Aug  3 22:34:44 UTC 2024 (`1722724484`) |
| **Workflow run** | [10231419317](https://github.com/storybookjs/storybook/actions/runs/10231419317) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=28800`_
</details>
<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  76.3 MB | 76.3 MB | 0 B | -0.11 | 0% |
| initSize |  167 MB | 167 MB | 379 B | -0.65 | 0% |
| diffSize |  91.1 MB | 91.1 MB | 379 B | -0.65 | 0% |
| buildSize |  7.42 MB | 7.42 MB | 0 B | **1.36** | 0% |
| buildSbAddonsSize |  1.61 MB | 1.61 MB | 0 B | **-1.36** | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  2.29 MB | 2.29 MB | 0 B | **1.36** | 0% |
| buildSbPreviewSize |  351 kB | 351 kB | 0 B | **-1.36** | 0% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  4.45 MB | 4.45 MB | 0 B | **1.36** | 0% |
| buildPreviewSize |  2.97 MB | 2.97 MB | 0 B | **-1.36** | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |



| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  21s | 17.8s | -3s -287ms | -0.27 | -18.5% |
| generateTime |  18.9s | 21.3s | 2.3s | -0.14 | 11.1% |
| initTime |  16s | 17.9s | 1.9s | -0.21 | 10.8% |
| buildTime |  15.2s | 11.8s | -3s -381ms | -0.34 | -28.6% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  6.6s | 6.8s | 249ms | **-1.49** | 3.6% |
| devManagerResponsive |  4.2s | 4.3s | 72ms | **-1.55** | 1.7% |
| devManagerHeaderVisible |  690ms | 671ms | -19ms | **-1.43** | -2.8% |
| devManagerIndexVisible |  717ms | 674ms | -43ms | **-1.71** | 🔰-6.4% |
| devStoryVisibleUncached |  1.2s | 1s | -164ms | **-1.35** | 🔰-15.7% |
| devStoryVisible |  736ms | 716ms | -20ms | **-1.46** | -2.8% |
| devAutodocsVisible |  607ms | 627ms | 20ms | -1.2 | 3.2% |
| devMDXVisible |  650ms | 637ms | -13ms | -1.11 | -2% |
| buildManagerHeaderVisible |  682ms | 640ms | -42ms | **-1.31** | 🔰-6.6% |
| buildManagerIndexVisible |  683ms | 643ms | -40ms | **-1.31** | 🔰-6.2% |
| buildStoryVisible |  722ms | 680ms | -42ms | **-1.32** | 🔰-6.2% |
| buildAutodocsVisible |  596ms | 599ms | 3ms | **-1.28** | 0.5% |
| buildMDXVisible |  565ms | 584ms | 19ms | -1.12 | 3.3% |

<!-- BENCHMARK_SECTION -->

<!-- greptile_comment -->

## Greptile Summary

The pull request introduces a new experimental `@storybook/nextjs-vite` package, integrating Vite with Next.js for Storybook.

- **Framework Support**: Added `nextjs-vite` to `framework-to-renderer.ts`, `versions.ts`, and `frameworks.ts` to map it to the `react` renderer.
- **Configuration Files**: Added `.eslintrc.json`, `README.md`, `package.json`, `preset.js`, and `project.json` for the `nextjs-vite` framework.
- **Mock Implementations**: Introduced mocks for Next.js cache, headers, navigation, and router functionalities in `src/export-mocks`.
- **Decorators and Providers**: Added decorators for head management, images, routing, and styled JSX in `src`.
- **TypeScript Types**: Defined new types for framework options and configurations in `types.ts` and `typings.d.ts`.

Ensure all dependencies and configurations are correctly set up and thoroughly tested.

<!-- /greptile_comment -->

valentinpalkovic committed:
Add @storybook/nextjs-vite package

valentinpalkovic committed:
Formatting

valentinpalkovic committed:
Add 'nextjs-vite' as a supported framework in helpers.ts

valentinpalkovic committed:
Fix check step

valentinpalkovic committed:
Cleanup

valentinpalkovic committed:
Upgrade vite-plugin-storybook-nextjs

valentinpalkovic committed:
Edit Next.js documentation

ndelangen commented on docs/get-started/frameworks/nextjs.mdx:
```diff
@@ -82,6 +82,32 @@ Storybook for Next.js is a [framework](../../contribute/framework.mdx) that make
 
   {/* prettier-ignore-end */}
 
+  #### @storybook/nextjs-vite - Experimental Vite support
+
+  You can use our freshly baked experimental `@storybook/nextjs-vite` framework, which is based on Vite and removes the bloat of Webpack and Babel.
+
+  {/* prettier-ignore-start */}
+
+  <CodeSnippets path="nextjs-vite-install.md" />
+
+  {/* prettier-ignore-end */}
+
+  Then, update your `.storybook/main.js|ts` to change the framework property:
+
+  {/* prettier-ignore-start */}
+
+  <CodeSnippets path="nextjs-vite-add-framework.md" />
+
+  {/* prettier-ignore-end */}
+
+  Finally, if you were using Storybook plugins to integrate with Next.js, those are no longer necessary when using this framework and can be removed:
+
+  {/* prettier-ignore-start */}
+
+  <CodeSnippets path="nextjs-vite-remove-addons.md" />
+
+  {/* prettier-ignore-end */}
+
```
docs changes should be vetted by @jonniebigodes ?

ndelangen commented on code/frameworks/nextjs-vite/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    // "module": "Preserve",
+    // "moduleResolution": "Bundler"
```
?

ndelangen commented on scripts/prepare/check.ts:
```diff
@@ -39,7 +39,10 @@ run({ cwd: process.cwd() }).catch((err: unknown) => {
 function getTSDiagnostics(program: ts.Program, cwd: string, host: ts.CompilerHost): any {
   return ts.formatDiagnosticsWithColorAndContext(
-    ts.getPreEmitDiagnostics(program).filter((d) => d.file.fileName.startsWith(cwd)),
+    ts.getPreEmitDiagnostics(program).filter((d) => {
+      // console.log(d);
+      return d.file.fileName.startsWith(cwd);
+    }),
```
?

valentinpalkovic committed:
Cleanup

kylegach committed:
Docs updates

- Use a more stable heading
    - "Experimental" is temporary
- Update package names

valentinpalkovic committed:
Rename nextjs-vite to experimental-nextjs-vite

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/nextjs-vite

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/nextjs-vite

valentinpalkovic committed:
Bump version of @storybook/experimental-nextjs-vite

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/nextjs-vite

valentinpalkovic committed:
Fix test

valentinpalkovic committed:
Use jiti instead of esbuild-register

kylegach committed:
Add webpackFinal -> viteFinal callout

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/nextjs-vite

valentinpalkovic committed:
Add experimental-nextjs-vite/default-ts to daily set of sandboxes

valentinpalkovic committed:
Merge branch 'next' into valentin/nextjs-vite

valentinpalkovic committed:
Lint fixes

valentinpalkovic committed:
Fix nextjs/prerelease sandbox
